### ISSUE TITLE
Overview of ZooKeeper Issues and Related

### ISSUE THREAD

**Issue by markrmiller** [markrmiller/solr#2]
A kind of broad initial overview of ZooKeeper challenges and issues and the many things that tie into that. More detail and various solutions and improvements to come in sub issues.

**Comment by markrmiller** [comment]
A handful of Videos to overview and outline will be pipelining into this issue today. Detail and break out and text will show up in the various sub and related issues that are in this category.

https://user-images.githubusercontent.com/448788/114565797-403ff600-9c37-11eb-823e-13eea6be3fb4.mp4

**Comment by markrmiller** [comment]
Starting to dig into Leader Election...

https://user-images.githubusercontent.com/448788/114571286-294fd280-9c3c-11eb-8d9b-21bc543f2db9.mp4

**Comment by markrmiller** [comment]
A few more coming on issue overview and then we will do one focused more on how things are supposed to be done.

**Comment by markrmiller** [comment]
https://user-images.githubusercontent.com/448788/114584224-0cb99780-9c48-11eb-82dc-74c924bac2fb.mp4

**Comment by markrmiller** [comment]
https://user-images.githubusercontent.com/448788/114595842-717aef00-9c54-11eb-8556-6d9b30aad405.mp4

**Comment by murblanc** [comment]
I do sympathize with your examples of unneeded complexity and misuse of ZK.
That's the same direction I'm trying to steer my work (that you might not have seen if your snapshot is almost a year old).
In current main I've already merged (a while ago) doing cluster state updates from any thread (i.e. from the Collection API threads essentially) without going through Overseer, and now I have a PR out for doing the Collection API calls from any thread (i.e. from the CollectionsHandler receiving the user requests).

These two aspects put together, then there's no longer a need for Overseer, and the bottom part of your class diagram goes away and/or is collapsed into the upper part.

And then (next items I plan to work on), stop watching all the state in the world all the time. I believe for most of our cluster management needs (and API handling), we do not need watches (we do need them when we send a request to another node to find out when that node is done, but this is pretty rare occurrence of creating a replica etc).

**Comment by markrmiller** [comment]
This largely wraps the quick spin through ZK - skips lots of stuff and other things I've tried and that you can do, but IMO ends pretty positive in terms of potential. The scale and efficiency and problem areas - so much of that can just be excised.

https://user-images.githubusercontent.com/448788/114607774-8b233300-9c62-11eb-9c0b-d95e78848933.mp4

**Comment by markrmiller** [comment]
> I do sympathize with your examples of unneeded complexity and misuse of ZK.
> That's the same direction I'm trying to steer my work (that you might not have seen if your snapshot is almost a year old).
> In current main I've already merged (a while ago) doing cluster state updates from any thread (i.e. from the Collection API threads essentially) without going through Overseer, and now I have a PR out for doing the Collection API calls from any thread (i.e. from the CollectionsHandler receiving the user requests).
> 
> These two aspects put together, then there's no longer a need for Overseer, and the bottom part of your class diagram goes away and/or is collapsed into the upper part.
> 
> And then (next items I plan to work on), stop watching all the state in the world all the time. I believe for most of our cluster management needs (and API handling), we do not need watches (we do need them when we send a request to another node to find out when that node is done, but this is pretty rare occurrence of creating a replica etc).

**Comment by murblanc** [comment]
I think I got the idea in your last video above: one persistent recursive watch getting all change notifications. I didn't get though how the notifications get to the other clients, running on other nodes, that want them (assuming we're trying to replace existing architecture, like all nodes having a replica for a given shard might be interested in changes to the shard). Do we maintain a parallel id hierarchy and the single watcher notifies through synthetic changes to that hierarchy (that the others are... watching?) or a different distribution mechanism that is not ZK based? Or something else?

**Comment by markrmiller** [comment]
I just hit wrong buttons, but corrected reply above.

**Comment by markrmiller** [comment]
I also should state, in any i divídale topic, improvement, feature, way of doing things - any smart devs, ransom devs, they will validly no better - be able to correct and improve on whatever I say or claim. I've got to spread information, that's a line that look distorted from the other side. 

Because the big issue to confront is the huge surface area and connection and interaction of things. So I've got to lay things out in a way that's possible - have to communicate, have to have a flow of information.  But what we are aiming towards is not any of these early or individual things. That's why it's crucial to take a little as I get it all out. Incorrect, insufficient, distorted, misremembered, misspoken or whatever. What we want to get to is not any of my lecture points - but the large survey results. They lay if the land in terms of how many things are working against good work. There is plenty of good work. Of things that I look at and say, yeah, nice move, likely would not have thought of it or done it. Some great people have come through. This is the tip of the iceberg that I want to start getting conveyed so that instead of me laying out all the troubles and results and barriers and crap I've run into and found blah blah, we can move to sensible collaboration and attack plans.
